src/lib.rs

```diff
@@ -253,16 +302,44 @@ pub fn gen_register_r(r: &Register, d: &Defaults) -> Vec<Tokens> {
                 }
             }
         } else {
-            let width_ty = width.to_ty();
+            let ty = field.to_ty();
             let mask: u64 = (1 << width) - 1;
             let mask = Lit::Int(mask, IntTy::Unsuffixed);
 
-            quote! {
-                pub fn #name(&self) -> #width_ty {
-                    const MASK: #bits_ty = #mask;
-                    const OFFSET: u8 = #offset;
+            if let Some(enum_vals) = field.enumerated_values.as_ref() {
+                let keys = enum_vals.values
+                    .iter()
+                    .map(|x| {
+                        Ident::new(ty.as_ref().to_owned() + "::" +
+                                   x.name.sanitize().to_pascal_case().as_ref())
+                    })
+                    .collect::<Vec<_>>()
+                    .into_iter();
+                let vals = enum_vals.values
+                    .iter()
+                    .map(|x| x.value)
+                    .collect::<Vec<_>>()
+                    .into_iter();
+
+                quote! {
+                    pub fn #name(&self) -> #ty {
+                        const MASK: #bits_ty = #mask;
+                        const OFFSET: u8 = #offset;
+
+                        match ((self.bits >> OFFSET) & MASK) {
+                            _ => ::core::option::Option::None,
+                            #(#vals => ::core::option::Option::Some(#keys)),*
+                        }.unwrap()
```

Comment:
`unwrap`? :scream_cat: 

This should return `Option` IMO. Most people will be using this like `reg.bit() == Some(Variant)` or via `if let Some(variant) = reg.bit()` rather than unwrapping what `reg.bit()` returns.